src/coreclr/src/vm/interoplibinterface.cpp

```diff
@@ -1452,6 +1453,90 @@ bool GlobalComWrappersForMarshalling::TryGetOrCreateObjectForComInstance(
     }
 }
 
+void QCALLTYPE GlobalComWrappersForTrackerSupport::SetGlobalInstanceRegisteredForTrackerSupport()
+{
+    // QCALL contracts are not used here because the managed declaration
+    // uses the SuppressGCTransition attribute
+
+    _ASSERTE(!g_IsGlobalComWrappersRegisteredForTrackerSupport);
+    g_IsGlobalComWrappersRegisteredForTrackerSupport = true;
+}
+
+bool GlobalComWrappersForTrackerSupport::TryGetOrCreateComInterfaceForObject(
+    _In_ OBJECTREF instance,
+    _Outptr_ void** wrapperRaw)
+{
+    if (!g_IsGlobalComWrappersRegisteredForTrackerSupport)
+        return false;
+
+    // Switch to Cooperative mode since object references
+    // are being manipulated.
+    {
+        GCX_COOP();
+
+        // Passing NULL as the ComWrappers implementation indicates using the globally registered instance
+        return TryGetOrCreateComInterfaceForObjectInternal(
+            NULL,
+            instance,
+            CreateComInterfaceFlags::CreateComInterfaceFlags_TrackerSupport,
+            ComWrappersScenario::TrackerSupportGlobalInstance,
+            wrapperRaw);
+    }
+}
+
+bool GlobalComWrappersForTrackerSupport::TryGetOrCreateObjectForComInstance(
+    _In_ IUnknown* externalComObject,
+    _Out_ OBJECTREF* objRef)
+{
+    if (!g_IsGlobalComWrappersRegisteredForTrackerSupport)
+        return false;
+
+    // Determine the true identity of the object
+    SafeComHolder<IUnknown> identity;
+    {
+        GCX_PREEMP();
+
+        HRESULT hr = externalComObject->QueryInterface(IID_IUnknown, &identity);
+        _ASSERTE(hr == S_OK);
+    }
+
+    // Switch to Cooperative mode since object references
+    // are being manipulated.
+    {
+        GCX_COOP();
+
+        // Passing NULL as the ComWrappers implementation indicates using the globally registered instance
+        return TryGetOrCreateObjectForComInstanceInternal(
+            NULL /*comWrappersImpl*/,
+            identity,
+            CreateObjectFlags::CreateObjectFlags_TrackerObject,
+            ComWrappersScenario::TrackerSupportGlobalInstance,
+            NULL /*wrapperMaybe*/,
+            objRef);
+    }
+}
+
+IUnknown* ComWrappersNative::GetIdentityForObject(_In_ OBJECTREF* objectPROTECTED)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+        PRECONDITION(CheckPointer(objectPROTECTED));
+    }
+    CONTRACTL_END;
+
+    ASSERT_PROTECTED(objectPROTECTED);
+
+    void* context;
+    if ((*objectPROTECTED)->GetSyncBlock()->GetInteropInfo()->TryGetExternalComObjectContext(&context))
```

Comment:
Much of the perf issue is in here. The entire `SyncBlock`/`InteropInfo` chain of operations are lazy allocation calls. 
* `GetSyncBlock()` should be `PassiveGetSyncBlock()` - Check for `null`. Unsure if this is warranted since the object should be valid. @jkotas Does this make sense? 
* `GetInteropInfo()` should be `GetInteropInfoNoCreate()` - Check for `null`.

If there is an `InteropInfo` only then should we try to call `TryGetExternalComObjectContext()`.

Reply:
Looks like calling `PassiveGetSyncBlock()` should also be done. I thought all valid objects had one automatically. I appear to be wrong. Check for null.